packages/timeline/src/sheets/build-nested-data.js

```diff
@@ -0,0 +1,166 @@
+  let _stacks = []
+  for (let i = 0; i < length; i++) {
+    _stacks.push([])
+  }
+  return _stacks
+}
+
+/**
+ *
+ *
+ * @param {string} elementType
+ * @returns {number} return the heading level of given element type.
+ *                   if the element is not a heading element, it will return -1
+ */
+function getElementHeadingLevel(elementType) {
+  return sectionLevels.findIndex(
+    level => _.get(level, 'heading.type') === elementType
+  )
+}
+
+/* 
+EXAMPLE 1
+  INPUT :
+    HEADING ELEMENT:
+      unitFlag9
+    STACKS:
+      [
+        [],                       // groups
+        [],                       // units
+        [record10, record11],     // records
+      ]
+  OUTPUT:
+    [
+      [],
+      [[unitFlag9, [record10, record11]]],
+      [],
+    ]
+
+EXAMPLE 2
+  INPUT :
+    HEADING ELEMENT:
+      groupFlag6
+    STACKS:
+      [
+        [],                                   // groups
+        [[unitFlag9, [record10, record11]]],  // units
+        [record7, record8],                   // records
+      ]
+  OUTPUT:
+    [
+      [
+        [
+          groupFlag6,
+          [
+            [null, [record7, record8]],
+            [unitFlag9, [record10, record11]]
+          ]
+        ],
+      ],
+      [],
+      [],
+    ]
+*/
+
+/**
+ * Build the `section` to which the heading element belong.
+ * The elements and the subsections in the stacks will be aggregate into the section.
+ *
+ * @param {Object} headingElement
+ * @param {[][]} stacks
+ * @returns {[]} a section array like [headingElement, [...subsectionsOrElements]]
+ */
+function aggregateSection(headingElement, stacks) {
+  const newStacks = buildEmptyStacks(levelsCount)
+  const targetLevel = getElementHeadingLevel(headingElement.type)
+  for (let level = maxLevel; level > targetLevel; level--) {
+    const _levelStack = stacks[level]
+    if (_levelStack.length > 0) {
+      const _headingElement = targetLevel === level - 1 ? headingElement : null
+      const _section = [_headingElement, [..._levelStack]]
+      newStacks[level - 1].unshift(_section)
+    }
+  }
+  return newStacks[targetLevel][0]
+}
+
+/**
+  Transform flat elements into nested sections.
+  A `section` can be represent by an array with two items: `[headingElement, [...subsectionsOrElements]]`
+
+    INPUT:
+          [
+            groupFlag1,
+            unitFlag2,
+            record3,
+            groupFlag4, 
+            record5,
+            record6,
+            unitFlag7,
+            record8,
+            unitFlag9,
+            record10,
+            record11,
+          ]
+    OUTPUT:
+          [
+            [
+              groupFlag1,
+              [
+                [unitFlag2, [record3]]
+              ]
+            ],
+            [
+              groupFlag4,
+              [
+                [null, [record5, record6]],
+                [unitFlag7, [record8]],
+                [unitFla9, [record10, record11]]
+              ]
+            ]
+          ]
```

Comment:
If we could add these comments,
that would be good for understanding the function.
```
const ele1 = {
  type: 'grou-flag',
  label: '2018',
}  // ele
const elel2 = {
  type: 'unit-flag',
  label: '2018.2.17',
}
const elel3 = {
  type: 'record',
  title: 'record title',
  description: 'This is a description of this record',
}

// INPUT
[ 
  ele1,
  ele2,
  ele3,
  ... // omit here due to explanation
]

// OUTPUT
[
  [
    ele1, // group 
    [
      [ele2 /*unit*/, [ele3]/*record*/]
    ]
  ]
  ... // omit here due to explanation
]
```